server/datastore/mysql/queries.go

```diff
@@ -40,115 +45,147 @@ func (ds *Datastore) ApplyQueries(ctx context.Context, authorID uint, queries []
+			return ds.withRetryTxx(ctx, func(tx sqlx.ExtContext) error {
+				// For upserting
+				pToInsert := make([]string, 0, len(batch))
+				aToInsert := make([]interface{}, 0, len(batch)*13)
+
+				// For fetching the ID after the upsert
+				pToSelect := make([]string, 0, len(batch))
+				aToSelect := make([]interface{}, 0, len(batch)*2)
+
+				for _, q := range batch {
+					pToInsert = append(pToInsert, "( ?, ?, ?, ?, true, ?, ?, ?, ?, ?, ?, ?, ?, ? )")
+					aToInsert = append(aToInsert, q.Name, q.Description, q.Query, authorID, q.ObserverCanRun, q.TeamID,
+						q.TeamIDStr(), q.Platform, q.MinOsqueryVersion, q.Interval, q.AutomationsEnabled, q.Logging,
+						q.DiscardData)
+
+					pToSelect = append(pToSelect, "(name = ? AND team_id_char = ?)")
+					aToSelect = append(aToSelect, q.Name, q.TeamIDStr())
 				}
+
+				upsertStm := fmt.Sprintf(upsertQueriesSQL, strings.Join(pToInsert, ","))
+				if _, err = tx.ExecContext(ctx, upsertStm, aToInsert...); err != nil {
+					return ctxerr.Wrap(ctx, err, "bulk upserting queries")
+				}
+
+				selectStm := fmt.Sprintf(
+					`SELECT id, name, team_id FROM queries WHERE %s`,
+					strings.Join(pToSelect, " OR "),
+				)
+				rows, err := tx.QueryContext(ctx, selectStm, aToSelect...)
 				if err != nil {
-					return ctxerr.Wrap(ctx, err, "select queries id")
+					return ctxerr.Wrap(ctx, err, "select queries for update")
 				}
-				// Get the ID from the rows
-				if rows.Next() {
-					if err := rows.Scan(&id); err != nil {
-						return ctxerr.Wrap(ctx, err, "scan queries id")
+				for rows.Next() {
+					var id uint
+					var name string
+					var teamID *uint
+					if err := rows.Scan(&id, &name, &teamID); err != nil {
+						return ctxerr.Wrap(ctx, err, "scan existing query")
+					}
+					if q, ok := batchGrp[unqKeyGen(name, teamID)]; ok {
+						q.ID = id
 					}
-				} else {
-					return ctxerr.Wrap(ctx, err, "could not find query after update")
 				}
-				if err = rows.Err(); err != nil {
-					return ctxerr.Wrap(ctx, err, "err queries id")
+				if err := rows.Err(); err != nil {
+					return ctxerr.Wrap(ctx, err, "fetching query IDs")
 				}
-				if err := rows.Close(); err != nil {
-					return ctxerr.Wrap(ctx, err, "close queries id")
+				if err := rows.Close(); err != nil { //nolint:sqlclosecheck
+					return ctxerr.Wrap(ctx, err, "closing query rows")
 				}
 
+				return ds.updateQueryLabelsInTx(ctx, batch, tx)
+			})
+		}
+
+		const maxRetries = 5
+		for attempt := 0; attempt < maxRetries; attempt++ {
+			err := processBatch()
+			if err == nil {
+				return nil
 			}
-			//nolint:gosec // dismiss G115
-			q.ID = uint(id)
 
-			err = ds.updateQueryLabelsInTx(ctx, q, tx)
-			if err != nil {
-				return ctxerr.Wrap(ctx, err, "exec queries update labels")
+			base := ctxerr.Cause(err)
+			var mysqlErr *mysql.MySQLError
+			if errors.As(base, &mysqlErr) {
+				if mysqlErr.Number != mysqlerr.ER_LOCK_DEADLOCK && mysqlErr.Number != mysqlerr.ER_LOCK_WAIT_TIMEOUT {
```

Comment:
`ds.withRetryTxx` already retries on these errors with exponential backoff. Not sure if double retrying makes sense.